File: JAX/linear_reg.py
import pandas as pd
import matplotlib.pyplot as plt
import jax.numpy as jnp
from jax import grad, vmap, jit
from jax import random
from sklearn.datasets import load_iris, load_diabetes


DARK_BLUE: str = '#1F77B4'


def laod_scikit_data(task: str = 'regression'):
    """Load dataset for the preferred task"""
    if task == 'classification':
        return load_iris(return_X_y=True, as_frame=True)
    elif task == 'regression':
        return load_diabetes(return_X_y=True, as_frame=True)


def plot_model(Y, Y_hat, w):
    fig = plt.figure(figsize=(10,8))
    plt.scatter(range(Y.shape[0]), Y)
    plt.scatter(range(Y.shape[0]), Y_hat)
    plt.legend(['True Values', 'Predictions'])
    plt.title('Linear Regression Model Prediction')
    plt.show()


def random_key_init(key: int =42):
    """Initializing random key for JAX library"""
    rnd_key = random.PRNGKey(key)
    return rnd_key


def add_bias_into_matrices(X):
    """Add initial column of ones to act as bias"""
    return jnp.insert(X, 0, 1, axis=1)


def linear_reg_pred(X, w):
    """Making prediction for each sample in the data"""
    return jnp.dot(X, w)


def mse_loss(X, Y, w):
    """Calculating the MSE loss using true values and prediction"""
    Y_hat = linear_reg_pred(X, w)
    return jnp.sum(Y - Y_hat)/Y.shape[0]


def linear_reg(X, Y):
    """Train linear regression model for given X and Y"""
    X_np = add_bias_into_matrices(X.values)  # Convert pd.DataFrame into np array
    # Convert labels to np array and add one more dim to mach sizes
    Y_np = jnp.expand_dims(Y.values, -1)
    w = jnp.dot(jnp.linalg.inv(jnp.dot(X_np.T,X_np)),jnp.dot(X_np.T,Y_np))
    mse = mse_loss(X_np, Y_np, w)  # Calculate the loss
    print('Linear Reg MSE error :', mse)
    Y_hat_linear_reg = linear_reg_pred(X_np, w)
    plot_model(Y, Y_hat_linear_reg, w)
    return Y_hat_linear_reg, w


if __name__ == '__main__':
    task = 'regression'
    X, Y = laod_scikit_data(task) 
    linear_reg(X, Y)

File: Julia/linear_reg.jl
using DataFrames
using Plots
using MLDatasets: BostonHousing, Iris


"""Loading datasets for the given tasks"""
function load_dataset(ml_task="regression")
    if ml_task == "regression"
        return BostonHousing.features()', BostonHousing.targets()'
    elseif ml_task == "classification"
        return Iris.features(), Iris.labels()
    end
end


"""Making linear regression prediction using data and weights"""
function linear_reg_pred(X, w)
    return X * w
end


"""MSE loss function"""
function mse_loss(X, y, w)
    y_hat = linear_reg_pred(X, w)
    return sum(y - y_hat) / size(y)[1]
end


"""Linear regression algortihm"""
function linear_reg(X, Y)
    X_with_bias = hcat(ones(size(X)[1],1),X)
    w = inv(X_with_bias' * X_with_bias) * (X_with_bias' * Y)
    mse = mse_loss(X_with_bias, y, w)
    println("Linear Regression MSE error : $mse")
    return linear_reg_pred(X_with_bias, w), w
end

gr()
X, y = load_dataset();
y_hat, weights = linear_reg(X, y);
plot(collect(1:size(y)[1]), y, seriestype = :scatter, title = "Linear Regression Result");
plot!(collect(1:size(y)[1]), y_hat, seriestype = :scatter)
